### PR TITLE
[EdgeDB] ProgressReport Workflow Queries

### PR DESCRIPTION
**Pull request by atGit2021**
[Monday](https://seed-company-squad.monday.com/boards/5989610236/views/131962501/pulses/5799644779)

### CODE REVIEW

**Comment by atGit2021** [comment]
> @atGit2021 this worked for me
> 
> ```ts
>   readonly transition?: InternalTransition & SetUnsecuredType<ID | null>;
> ```

Thank you

File: src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    hydrate: (event) => ({
+      ...event['*'],
+      report: true,
+      who: true,
+      transition: event.transitionId,
+    }),
+  }).customize((cls, { defaults }) => {
+    return class extends cls {
+      static omit = [defaults.create, defaults.update, defaults.delete];
+      async list(
```
Comment (bryanjnelson) [inline src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:28]: @CarsonF - Andre's reasoning for customizing this is because the signature in the common EdgeDB repo uses Pagination, but the one in the Neo4j repo does not.  Is customizing the best approach here?  Should we consider altering the Neo4j list instead?

Comment (CarsonF) [reply src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:28]: The list isn't paginated in GQL, so we couldn't just change the neo4j query.
https://github.com/SeedCompany/cord-api-v3/blob/develop/src/components/progress-report/workflow/resolvers/progress-report-workflow-events.resolver.ts#L11-L17

Customizing the method here is the right approach.

File: src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            id: event.who.id,
+            email: event.who.email,
+          },
+        }));
+      }
+    };
+  })
+  implements PublicOf<ProgressReportWorkflowRepository>
+{
+  async recordTransition(
```
Comment (bryanjnelson) [inline src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:72]: @CarsonF - Would it make sense to move both `recordTransition` and `recordBypass` up to the service so that they don't need to be duplicated here?  Neither of them are actually doing anything database specific.

Comment (CarsonF) [reply src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:72]: Sure

**Review by bryanjnelson — CHANGES_REQUESTED** [review]
Excellent work, @atGit2021!

File: src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    const users = await this.db.run(query);
+    return users.map((user) => ({
+      id: user.id,
+      email: user.email,
+      roles: user.roles.map((role) => role),
+    }));
+  }
+
+  async getUserIdByEmails(emails: readonly string[]) {
+    const query = e.params({ emails: e.array(e.str) }, ({ emails }) =>
```
[the remark this answers is not in the document]
Comment (CarsonF) [reply src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:73]: Params is better because it doesn't change the edgeql query with every different input. This allows the server to cache it.

File: src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    const query = e.select(
+      e.select(progressReport).project.members.user,
+      () => ({
+        id: true,
+        email: true,
+        roles: true,
+      }),
+    );
+    const users = await this.db.run(query);
+    return users.map((user) => ({
```
[the remark this answers is not in the document]
Comment (atGit2021) [reply src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:65]: I get the following error, `Types of property 'roles' are incompatible.` when I don't map through the roles, e.g. ` roles: user.roles.map((role) => role),`

File: src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bryanjnelson) [reply src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:73]: Ah.  Well, that makes sense.  It's unfortunate, but it makes sense.

File: src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts [hunk 5]
```diff
@@ -76,7 +76,7 @@ export class ProgressReportWorkflowNotificationHandler
         this.prepareNotificationObject(
           reportId,
           previousStatus,
-          userId ? { userId } : { email },
+          userId ? { userId } : email ? { email } : { email: '' },
```
Comment (CarsonF) [inline src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts:79]: Empty strings to hide type errors are a big ick for me.
Instead, in `getProjectNotifyees`, member's without email addresses can be `flatMap`'ped out.

**Review by CarsonF — CHANGES_REQUESTED** [review]
Thanks @atGit2021! Just a couple things.

File: src/components/progress-report/workflow/progress-report-workflow.repository.ts [hunk 6]
```diff
@@ -81,33 +79,7 @@ export class ProgressReportWorkflowRepository extends DtoRepository(
...
-  ) {
-    return await this.recordEvent(report, { status, notes }, session);
-  }
-
-  private async recordEvent(
-    report: ID,
-    props: Record<string, any>,
-    session: Session,
-  ) {
+  async recordEvent(report: ID, props: Record<string, any>, session: Session) {
```
[the remark this answers is not in the document]
Comment (atGit2021) [reply src/components/progress-report/workflow/progress-report-workflow.repository.ts:82]: @CarsonF using the `ExecuteProgressReportTransitionInput ` type allows for 'undefined' to be passed in to props.status which is not in the ProgressReportStatus Enum.  Should I coalesce the value to "Not Started" if undefined is received or should I code a guard to not update that property if undefined is received?

[the remark this answers is not in the document]
Comment (CarsonF) [reply src/components/progress-report/workflow/progress-report-workflow.repository.ts:82]: Ah. You can just adjust it to be required.
```ts
SetRequired<ExecuteProgressReportTransitionInput, 'status'>
```

[the remark this answers is not in the document]
Comment (atGit2021) [reply src/components/progress-report/workflow/progress-report-workflow.repository.ts:82]: @CarsonF I modified recordEvent to use the new type but I'm getting an error, see below.  I'm not sure how to make the types on both the edge and neo4j `recordEvent` method match.

Property 'recordEvent' in type 'ProgressReportWorkflowEdgeDBRepository' is not assignable to the same property in base type 'PublicOf<ProgressReportWorkflowRepository>'.
  Type '({ report, ...props }: { readonly report: ID; readonly transition?: ID | undefined; readonly notes?: RichTextDocument | undefined; readonly status: "NotStarted" | "InProgress" | "PendingTranslation" | "InReview" | "Approved" | "Published"; }, _session: Session) => Promise<...>' is not assignable to type '(report: ID, props: Record<string, any>, session: Session) => Promise<UnsecuredDto<ProgressReportWorkflowEvent>>'.
    Types of parameters '__0' and 'report' are incompatible.

File: src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts (hunk 5 above)
Comment (atGit2021) [reply src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts]: @CarsonF , I'm having trouble getting rid of the null email values. I pushed up a version using flatMap in the `getProjectNotifyees` method, so you can see the code.

File: src/components/progress-report/workflow/progress-report-workflow.repository.ts [hunk 7]
```diff
@@ -216,7 +173,7 @@ export class ProgressReportWorkflowRepository extends DtoRepository(
       .where({ 'email.value': inArray(emails) })
       .return<{
         id: ID;
-        email: string;
+        email: string | null;
```
Comment (CarsonF) [inline src/components/progress-report/workflow/progress-report-workflow.repository.ts:176]: You can revert this change. We know the emails exist, that's our input/starting point.

File: src/components/progress-report/workflow/progress-report-workflow.repository.ts [hunk 8]
```diff
@@ -180,8 +152,8 @@ export class ProgressReportWorkflowRepository extends DtoRepository(
       ])
       .return<{
         id: ID;
-        email: string;
-        roles: Role[];
+        email: string | null;
```
Comment (CarsonF) [inline src/components/progress-report/workflow/progress-report-workflow.repository.ts:155]: Let's revert this type change, and change the optional match on the email address to a required one.
Let the DB do the filtering for us.

File: src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts (hunk 5 above)
Comment (CarsonF) [reply src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts]: I couldn't get the flatMap to work with types either, not exactly sure why.
In the end, it's better to just to align the DB queries closer to our need.
See my other comments in the repo.

File: src/components/progress-report/workflow/progress-report-workflow.repository.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (CarsonF) [reply src/components/progress-report/workflow/progress-report-workflow.repository.ts:82]: You updated the signature in the edgdb repository but not the neo4j repository. They need to match.

**Review by CarsonF — CHANGES_REQUESTED** [review]

File: src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts [hunk 9]
```diff
@@ -57,37 +57,39 @@ export class ProgressReportWorkflowEdgeDBRepository
...
+      with emails := array_unpack(<array<str>>$emails),
+      users := (select User filter .email in emails)
+      select users {
+        id,
+        email := assert_exists(.email),
+      } 
+    `);
+    return await this.db.run(query, { emails });
   }
+  // async getUserIdByEmails(emails: readonly string[]) {
```
Comment (bryanjnelson) [inline src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:79]: Probably just forgot to delete this?

Comment (atGit2021) [reply src/components/progress-report/workflow/progress-report-workflow.edgedb.repository.ts:79]: oops, forgot to save the file after deleting it.  will push another commit now

**Review by CarsonF — APPROVED** [review]
Looks good! Thanks @atGit2021
